magpie/api/login/login.py

```diff
@@ -271,69 +279,127 @@ def login_success_external(request, external_user_name, external_id, email, prov
-            # provide the token user data, let the external provider update it on login afterwards
-            token_type, access_token = request.headers.get("Authorization").split()
-            data = {"access_token": access_token, "token_type": token_type}
-            cred = Credentials(authomatic_handler.config, token=access_token, token_type=token_type, provider=provider)
-            provider.credentials = cred
-            result = LoginResult(provider)
-            # pylint: disable=W0212
-            result.provider.user = result.provider._update_or_create_user(data, credentials=cred)  # noqa: W0212
-
-        # otherwise, use the standard login procedure
-        else:
-            result = authomatic_handler.login(WebObAdapter(request, response), provider_name)
-            if result is None:
-                if response.location is not None:
-                    return HTTPTemporaryRedirect(location=response.location, headers=response.headers)
-                return response
-
-        if result:
-            if result.error:
-                # Login procedure finished with an error.
-                error = result.error.to_dict() if hasattr(result.error, "to_dict") else result.error
-                LOGGER.debug("Login failure with error. [%r]", error)
-                return login_failure_view(request, reason=result.error.message)
-            if result.user:
-                # OAuth 2.0 and OAuth 1.0a provide only limited user data on login,
-                # update the user to get more info.
-                if not (result.user.name and result.user.id):
-                    try:
-                        response = result.user.update()
-                    # this error can happen if providing incorrectly formed authorization header
-                    except OAuth2Error as exc:
-                        LOGGER.debug("Login failure with Authorization header.")
-                        ax.raise_http(http_error=HTTPBadRequest, content={"reason": str(exc.message)},
-                                      detail=s.ProviderSignin_GET_BadRequestResponseSchema.description)
-                    # verify that the update procedure succeeded with provided token
-                    if 400 <= response.status < 500:
-                        LOGGER.debug("Login failure with invalid token.")
+        if provider_name == MAGPIE_NETWORK_PROVIDER:
+            if "Authorization" in request.headers:
+                token_type, token = request.headers.get("Authorization").split()
+                if token_type != "Bearer":
+                    ax.raise_http(http_error=HTTPBadRequest,
+                                  detail=s.ProviderSignin_GET_BadRequestResponseSchema.description)
+                authenticated_user = None
+                try:
+                    authenticated_user = user_from_token(request, token)
+                except jwt.exceptions.InvalidIssuerError:
```

Comment:
Perform error handling closer to the part where it actually happens. Return `None` and handle accordingly here.

Reply:
The `user_from_token` function is also used in `validate_token_view`. These functions handle the exceptions differently depending on which `jwt.exceptions` class is raised. If I handle it in the `user_from_token` view and then return `None` then the calling functions are not able to handle different errors differently.

Reply:
Ok.
The "how" to return the error needs to be given more though than just `None`. 
However, this is why I would prefer the try/except closer to the call that can raise it (I assume from `jwt.decode`). 
If `user_from_token` got refactored to handle exceptions differently, we might not notice that it break how it is used elsewhere. At least, I didn't notice until you pointed it out.